Prevent overlapping drop placements

e-kolpakov opened the pull request:
**Description:** This PR updates item placement logic:

* Zone alignment "none" (aka no alignment) is removed, dropped items always use the "aligned" layout option rather than being placed where the learner dropped them.
* "Max items per zone" setting added to allow limiting max number of items dropped into a single zone - course authors can use this setting to provide some kind of a hint, or to prevent visual overflow when too many items are dropped into single zone.

**JIRA:** [SOL-1979](https://openedx.atlassian.net/browse/SOL-1979)

**Discussions:** [SOL-1979](https://openedx.atlassian.net/browse/SOL-1979) and https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/83#issuecomment-236017205

**Sandboxes:**

* LMS: http://dndv2-sandbox9.opencraft.hosting/
* Studio: http://studio-dndv2-sandbox9.opencraft.hosting/


**Note:** at the moment #86 was merged, but #87 was not, and this PR does not include changes from #87; as a result, DnDv2 on sandbox show "Submit" button, but it does not do anything - this is an expected behavior until #87 is merged

**Testing instructions:**

Items alignment:
1. Open new or existing DnDv2 block in studio, open editor, go to "Zones" tab. *Expected:* zone "Alignment" dropdown does not have "none" option; default is "center"
2. Open DnDv2 XBlock in LMS.
3. Make sure that items dropped into a zone are aligned according to zone alignment configuration.
4. Existing zones with "none" alignment should be treated as having "center"alignment; this should be the case even if an XBlock was not updated in Studio after this PR is applied.

Max items per zone:
1. Open new or existing DnDv2 block in studio, open editor, go to "Items" tab. *Expected:* "Max items per zone" is displayed.
2. Setting "max items per zone" to ~~zero or~~ empty string has no further effect (make sure you can put as much items into a single dropzone as you wish in LMS)
3. Setting "max items per zone" to zero or negative value ~~should cause validation errors~~ will result in value being set to empty string
4. ~~Setting "max items per zone" to positive value will validate item placement. If any zone has more "correct" items than "max items per zone" a validation error will be shown, and XBlock will not be updated.~~
5. After setting "max items per zone" to positive value ~~and passing the validation~~, publish the unit and open it in LMS. *Expected:* can only put up to "max items per zone" items into a single zone.
6. *Note:* existing answers are not updated, so it is possible to have more than "max items per zone" items dropped into a single zone if "max items" setting changed after the answer was submitted.

**Screenshots:**

Studio "Zones" tab:

![image](https://cloud.githubusercontent.com/assets/3330048/17551549/79ba91d6-5f03-11e6-93ce-72042a44b884.png)

Studio "Items" tab:

![image](https://cloud.githubusercontent.com/assets/3330048/17770812/67533cc8-6548-11e6-8aa6-c996b833ee53.png)

<!-- With validation errors: -->

<!-- ![image](https://cloud.githubusercontent.com/assets/3330048/17782607/1b27990e-657d-11e6-8d1f-07b01b2caef1.png) -->

LMS - item alignment:

![image](https://cloud.githubusercontent.com/assets/3330048/17551802/d18e97d0-5f04-11e6-8ed2-467a9352ed05.png)

Attempting to exceed max items per zone:

![image](https://cloud.githubusercontent.com/assets/3330048/17551840/06e31b68-5f05-11e6-9f80-e39ae9363337.png)

**Reviewers:**
- [x] @itsjeyd 
- [x] TNL: @cahrens and/or @staubina
- [ ] ~~a11y: @cptvitamin~~  a11y review will be performed later, for all changes in v2.1: [SOL-1978](https://openedx.atlassian.net/browse/SOL-1978)
- [x] Product: @sstack22

**Checklist:**
- [x] Internal review
- [x] Upstream review
- [x] Squash commits

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx-solutions%2Fxblock-drag-and-drop-v2&number=92

itsjeyd commented on drag_and_drop_v2/default_data.py:
```diff
@@ -46,6 +47,8 @@
       "width": 340,
       "height": 138,
+      "align": "center"
+
```
@e-kolpakov Nit: Remove extra blank line

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -280,12 +299,60 @@ def studio_submit(self, submissions, suffix=''):
         }
 
+    @staticmethod
+    def _get_max_items(submission):
```
@e-kolpakov For consistency, can we rename `submission` to `submissions` here? When I first read this I found myself double-checking if `studio_submit` was passing the `submissions` it receives unaltered, or if it was passing only a subset of the `submissions` data.

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -280,12 +299,60 @@ def studio_submit(self, submissions, suffix=''):
+
+        Does not catch ValueError intentionally to allow validation to capture this problem
+        """
+        raw_max_submissions = submission.get('max_items_per_zone', None)
```
@e-kolpakov It would make sense to rename this to `raw_max_items` I think.

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -280,12 +299,60 @@ def studio_submit(self, submissions, suffix=''):
+
+        items = submissions['data']['items']
+
+        max_submissions = 0
```
@e-kolpakov Similar note here: `max_items` or `max_items_per_zone` would probably be a little more accurate.

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -280,12 +299,60 @@ def studio_submit(self, submissions, suffix=''):
+        if max_submissions:
+            counter = Counter()
+            for item in items:
+                zones = item.get('zones', [item.get('zone')])
```
@e-kolpakov Add an inline comment about backward compatibility here, perhaps?

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -444,10 +509,7 @@ def _get_item_state(self):
-                state[item_id] = item
-            else:
-                state[item_id] = {'top': item[0], 'left': item[1]}
+            state[item_id] = StateMigration.apply_item_state_migrations(item)
```
@e-kolpakov Thanks for factoring the migration logic out to utils.py! Makes the code much more readable +1

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -186,16 +174,13 @@ function DragAndDropTemplates(configuration) {
-            var is_item_in_zone = function(i) { return i.is_placed && (i.zone === zone.uid); };
-            items_in_zone = $.grep(ctx.items, is_item_in_zone);
-        }
+        item_wrapper += '.item-align.item-align-' + zone.align;
```
@e-kolpakov A couple nits:

* Maybe define `item_wrapper` in a single line
* Update the comment above to reflect the fact that all zones use alignment now

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -657,43 +641,48 @@ function DragAndDropBlock(runtime, element, configuration) {
-        var x_pos_percent = x_pos / $target_img.width() * 100;
-        var y_pos_percent = y_pos / $target_img.height() * 100;
+        var items_in_zone_count = countItemsInZone(zone);
+        if (configuration.max_items_per_zone && configuration.max_items_per_zone <= items_in_zone_count) {
```
@e-kolpakov If `max_items_per_zone` is equal to `items_in_zone_count`, that should probably be OK? In other words, should this code only reject the placement of an item if `items_in_zone_count > max_items_per_zone`?

e-kolpakov replied:
@itsjeyd no, this code works before the item is placed, so if we allow it to continue when `items_in_zone_count == max_items_per_zone` it will place one more item and we will exceed the limit.

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -657,43 +641,48 @@ function DragAndDropBlock(runtime, element, configuration) {
+    var countItemsInZone = function(zone) {
+        var count = 0;
+        for (var key in state.items) {
+            if (state.items.hasOwnProperty(key)) {
```
@e-kolpakov Why is this check necessary? (genuine question)

e-kolpakov replied:
@itsjeyd `state.items` is an object, and iterating over object this way yields all "own" properties and all "inherited" properties (properties from prototype), mostly functions, liek `toString` and so on.

I believe the approach we used here was to use `Object.keys(object).forEach(...)` - I just have missed that and written it old-style :)

itsjeyd replied:
@e-kolpakov I see, that makes sense. Thanks!

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -1024,32 +1010,8 @@ function DragAndDropBlock(runtime, element, configuration) {
-                delete item.absolute;
-            }
-        });
+    var migrateState = function() {
```
@e-kolpakov Leftover stub

e-kolpakov replied:
@itsjeyd no, it was not empty before. I thought it might make sense to keep it to give a hint to future developers (self?) that we can apply some migrations in JS.

itsjeyd replied:
@e-kolpakov OK, that's a bit unusual but I get where you're coming from. If you're not going to remove the method, could you update the docstring above to mention why it's there (and remove the part about image size)?

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop_edit.js:
```diff
@@ -514,18 +514,27 @@ function DragAndDropEditBlock(runtime, element, params) {
                         $('.xblock-editor-error-message', element).css('display', 'none');
                         var handlerUrl = runtime.handlerUrl(element, 'studio_submit');
+
+                        $('.xblock-editor-error-message', element).css('display', 'none');
```
@e-kolpakov Duplicate of L522. Could L521 be removed?

e-kolpakov replied:
@itsjeyd yes, nice catch - L521 and L522 totally filtered out from my attention; actually, we don't need this two lines added.

itsjeyd commented on drag_and_drop_v2/utils.py:
```diff
@@ -5,3 +5,135 @@
+
+
+class Constants(object):
+    ALLOWED_ZONE_ALIGNMENT = ['left', 'right', 'center']
```
@e-kolpakov Nit: `ALLOWED_ZONE_ALIGNMENTS` (since it's a list)

itsjeyd commented on drag_and_drop_v2/utils.py:
```diff
@@ -5,3 +5,135 @@
+        Migrates zone data from v1.0 format to v2.0 format.
+
+        Changes:
+        * v1 used zone "title" as UID, while v2 have "uid" property
```
@e-kolpakov `while v2 zones have a "uid" property`

itsjeyd commented on drag_and_drop_v2/utils.py:
```diff
@@ -5,3 +5,135 @@
+    @classmethod
+    def _zone_v2_to_v2p1(cls, zone):
+        """
+        Migrates zone data from v2.0 to v2.5
```
@e-kolpakov I think this should be `v2.0` to `v.2.1`. Same comment applies to the docstring of `_item_state_v1p5_to_v2p1` below.

itsjeyd commented on drag_and_drop_v2/utils.py:
```diff
@@ -5,3 +5,135 @@
+        Migrates item_state from v1.5 to v2.0
+
+        Changes:
+        * Item placement attributes switched from absolute (left-top) to relative (x_percent-y_pecrent) units
```
@e-kolpakov Typo: `pecrent`

itsjeyd commented on drag_and_drop_v2/utils.py:
```diff
@@ -5,3 +5,135 @@
+        Out: {'zone': 'Zone", 'correct': True, 'top': '100px', 'left': '120px'}
+        """
+        # Conversion can't be made as parent dimensions are unknown to python - converted in JS
+        # Since 2.5 JS this conversion became unnecesary, so it was removed from JS code
```
@e-kolpakov `v2.1`?

itsjeyd commented on drag_and_drop_v2/utils.py:
```diff
@@ -5,3 +5,135 @@
+        Migrates item_state from v1.5 to v2.5
+
+        Changes:
+        * Removed "none" zone alignment - remove old "absolute" placement attributes
```
@e-kolpakov Maybe turn this into two items (putting note about "absolute" placement first to match the examples below):

```
* Removed old "absolute" placement attributes
* Removed "none" zone alignment, making "x_percent" and "y_percent" attributes obsolete
```

itsjeyd commented on tests/integration/test_interaction.py:
```diff
@@ -817,28 +821,9 @@ def _assert_zone_align_item(self, item_id, zone_id, align, action_key=None):
     @data(
-        ([3, 4, 5], "Zone Invalid Align", "start"),
+        ([0, 1, 2], "Zone No Align", "center"),
+        ([3, 4, 5], "Zone Invalid Align", "center"),
```
@e-kolpakov Why is `"center"` an invalid alignment?

e-kolpakov replied:
@itsjeyd "Zone Invalid Align" refers to a zone that have `"align": "invalid"` in configuration. In this case, we expect items placed into that zone to be centered.

itsjeyd commented on tests/unit/data/old/config_out.json:
```diff
@@ -20,15 +21,17 @@
           "height": 100,
-          "uid": "Zone 1"
+          "uid": "Zone 1",
+          "align": "center"
```
@e-kolpakov Since this is supposed to represent v1 data, I'm wondering if we should remove the `"align"` setting here, and modify the migration code to add the setting if it's not present.

(The same goes for the `"uid"` setting, but I know you didn't add that.)

e-kolpakov replied:
@itsjeyd nope, the thing is, this particular test checks how DnDv2 parses old data format (aka v1) and how it responds to various fields/data members missing/having different names.

`config_out` files represent DnDv2 response to get_configuration handler as it is passed to frontend, so we expect it to have a most recent format (in this case v2.1)

itsjeyd replied:
@e-kolpakov Ah, right -- I thought I was commenting on one of the `data.json` files :) Sorry for the noise.

itsjeyd commented on tests/unit/test_advanced.py:
```diff
@@ -57,11 +57,16 @@ def initial_feedback(cls):
+        except AssertionError:
+            print self.expected_configuration()
+            print self.block.get_configuration()
+            raise
```
@e-kolpakov This can be useful, but maybe it's a leftover from testing on your local machine? :)

e-kolpakov replied:
@itsjeyd both - it is a leftover, but it is a useful leftover, because diffs are very hard to read, so most of the time it is simpler if you have those dicts printed.

CC @mtyaka @bradenmacdonald what do you think about this?

mtyaka replied:
This is very useful while debugging a problem, but I think I'd prefer to take it out and manually add it back in while debugging a failing test.

bradenmacdonald replied:
The problem is that committed code should almost never contain "print", unless you're writing some kind of interactive shell script. So if you leave it in, you should use this syntax:

```python
expected = self.expected_configuration()
actual = self.block.get_configuration()
self.assertEqual(
    expected,
    actual,
    "Unexpected configuration - Expected:\n{}\nObserved:\n{}".format(expected, actual)
)
```

p.s. You should get out of the habit of writing `print "foo"` since it's not python3 compatible :-p

itsjeyd commented on tests/unit/test_basics.py:
```diff
@@ -15,6 +15,29 @@ def setUp(self):
         self.block = make_block()
         self.patch_workbench()
 
+    def _make_submission(self, modify_submitssion=None):
```
@e-kolpakov Typo: `submitssion`

itsjeyd commented on tests/unit/test_basics.py:
```diff
@@ -107,33 +131,101 @@ def assert_user_state_empty():
+        self.assertEqual(self.block.max_items_per_zone, None)
+        self.assertEqual(self.block.data, {'foo': 1, 'items': []})
+
+    def test_studio_submit2(self):
```
@e-kolpakov Call this `test_studio_submit_assessment` or something like that?

e-kolpakov commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -708,6 +690,7 @@ function DragAndDropBlock(runtime, element, configuration) {
                         releaseItem($selectedItem);
                     } else if (isActionKey(evt)) {
                         evt.preventDefault();
+                        evt.stopPropagation();
```
@itsjeyd just to make sure - without this `stopPropagation` dropping an item using keyboard closed error popup before user could see it, resulting in item being returned to item bank without any explanation why.

itsjeyd replied:
@e-kolpakov Thanks for the explanation, I had wondered about that briefly but somehow forgot to comment about it - you read my mind :)

e-kolpakov commented on drag_and_drop_v2/public/js/drag_and_drop_edit.js:
```diff
@@ -161,7 +161,7 @@ function DragAndDropEditBlock(runtime, element, params) {
                             $('.save-button', element).parent()
                                 .removeClass('hidden')
-                                .one('click', function submitForm(e) {
+                                .on('click', function submitForm(e) {
```
@itsjeyd FYI important one-letter change: with `one` if submission failed due to validation error, `Save` button stop working, so you've got to cancel and go all over again.

itsjeyd replied:
@e-kolpakov OK. But now that I'm looking at the surrounding code I'm wondering what the purpose of [this code](https://github.com/open-craft/xblock-drag-and-drop-v2/blob/c1941da6629d38d41b5b2cc8b162b323dc5f80d6/drag_and_drop_v2/public/js/drag_and_drop_edit.js#L168-L171) is. It looks like it was supposed to work around the behavior you're describing. If that's the case, we should probably remove it -- the `.on` solution seems cleaner anyway.

bradenmacdonald commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -657,43 +640,42 @@ function DragAndDropBlock(runtime, element, configuration) {
+        var items_in_zone_count = countItemsInZone(zone);
+        if (configuration.max_items_per_zone && configuration.max_items_per_zone <= items_in_zone_count) {
+            state.last_action_correct = false;
+            state.feedback = gettext("This zone already contains maximum allowed number of items or more");
```
Better phrasing: "You cannot add any more items to this zone."

bradenmacdonald commented:
In this screenshot, the left-aligned items retain a nice small automatic size, but the centered ones grow to fill their zone width - is there a way to use the "left" sizing behavior in both cases?

![image](https://cloud.githubusercontent.com/assets/3330048/17551802/d18e97d0-5f04-11e6-8ed2-467a9352ed05.png)

Also, on the sandbox I wasn't able to see the warning in red when dragging too many items onto a zone. 

Finally, can you please enable the XBlock view on the sandbox so we can test how this works on small devices / in the mobile app?

http://studio-dndv2-sandbox9.opencraft.hosting/xblock/block-v1:OpenCraft+DnDv2+demos+type@vertical+block@28ae4b11831b42949d8815c86671424b should work but seems to need `FEATURES["ENABLE_XBLOCK_VIEW_ENDPOINT"]` set to True.

itsjeyd commented on tests/unit/test_basics.py:
```diff
@@ -107,33 +134,115 @@ def assert_user_state_empty():
+        def submission_success(submission):
+            submission['max_items_per_zone'] = 1
+            submission['data']['items'] = [
+                {'zones': ['1'], 'title': 'item 1'}, {'zones': ['2'], 'title': 'item 2'}
```
@e-kolpakov For consistency with the other submission modification methods, should zones be called `'Zone 1'` and `'Zone 2'` here as well?

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -280,12 +299,71 @@ def studio_submit(self, submissions, suffix=''):
+        errors = []
+        counter = Counter()
+        for item in items:
+            for zone in item.get('zones', []):
```
@e-kolpakov This doesn't need special logic for backward compatibility because it operates on item data from `submissions` passed to `studio_submit`, right? If it were operating on existing instance data, the following scenario might cause "item zone validation" to be skipped:

A course author sets `max_items_per_zone` to a positive integer on an old DnDv2 instance with items that still use `item["zone"]` instead of `item["zones"]` and then tries to save it for the first time. Since validation happens before any of the updated data gets saved, `item.get('zones', [])` would return `[]` for each item. As a result, the item count for each zone would be `0`, and the DnDv2 instance would pass validation even if there were more than `max_items_per_zone` items associated with any of the zones.

e-kolpakov replied:
@itsjeyd I don't think an old DnDv2 still using `item['zone']` will have this validation code - not like we're separately deploying backend and frontend here, so when this validation is added, JS will use `item["zones"]` - [it is already in master](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/630a8d7f519409c6c7237660557823998d9e9c29/drag_and_drop_v2/public/js/drag_and_drop_edit.js#L485-L487), while this validation is not.

e-kolpakov commented:
Rebased from https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92/commits/c1941da6629d38d41b5b2cc8b162b323dc5f80d6

e-kolpakov commented:
@itsjeyd test failures are from flaky tests again - could you please take a look at this PR again?

itsjeyd commented:
@e-kolpakov Testing on the sandbox right now :) What version/commit is the sandbox currently using?

itsjeyd commented:
@e-kolpakov Tested on the sandbox and everything works as advertised. The only thing that needs a couple of small adjustments is the "Maximum items per zone" setting:

* Adding a help text to the Studio editor would be nice. Something like "Upper limit for the number of items that can be associated with, and placed in, a zone. Leave unset if you don't want to constrain the number of items per zone."
* It would be good to have only a single value represent "unlimited items per zone". To be consistent with CAPA problems and other fields that set limits for certain things (e.g. the `max_attempts` field), that should be the empty string/None (and the `min` value for the input field should be `1`). That would open up the possibility for introducing "decoy" zones that don't take any items; but as long as we don't support that feature, manually entering `0` should cause a validation error, just like negative values do.

---

I also looked at your recent updates to the code, they're looking good. The only thing left to address is [this comment](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92/files#r74302172).

---

Finally, one question about the test instructions: Which unit on the sandbox should I use to verify this item:

> 4. Existing zones with "none" alignment should be treated as having "center"alignment; this should be the case even if an XBlock was not updated in Studio after this PR is applied.

e-kolpakov commented:
@itsjeyd Addressed your comments, but I don't have time to update the sandbox right now (read: until Monday). So, if you to test this before that - could you please update the sandbox? :) Thank you!

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -287,12 +306,76 @@ def studio_submit(self, submissions, suffix=''):
+        raw_max_items_per_zone = submissions.get('max_items_per_zone', None)
+
+        # tricky condition here: we want None and empty string considered "no value", but integer 0 as "value"
+        # Thus, can't use just `not raw_max_items_per_zone` - integer 0 will be converted to None
```
@e-kolpakov Did you mean "integer 0 would be converted to False" here?

itsjeyd commented on tests/unit/test_basics.py:
```diff
@@ -110,33 +137,127 @@ def assert_user_state_empty():
+        self.assertIn("Zone 1", res['messages'][0])
+
+    def test_studio_submit_max_items_validation_failure_multiple_items(self):
+        def submission_failure2(submission):
```
@e-kolpakov This should be called `modify_submission` to be consistent with the other functions that change the default submission. There's also a couple of functions that serve the same purpose and are called `update_submission` (in L157 and L185 above); it would make sense to rename those as well.

itsjeyd commented:
@e-kolpakov I updated the sandbox (you won't need to update it again after addressing the last remaining comments since they don't concern the application code) and verified the changes to the "Maximum items per zone" setting. You didn't make the help text for that field visible in the UI, but I'm OK with skipping that for now/leaving the decision to upstream.

The sandbox is now at https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92/commits/0270b0b035d88a350b814ddc6006f3186b3067ad.

So this is *almost* ready to go -- I'll just need to verify this item:

> Existing zones with "none" alignment should be treated as having "center"alignment; this should be the case even if an XBlock was not updated in Studio after this PR is applied.

Which unit on the sandbox should I use to do that? Is there any particular DnDv2 instance that you didn't modify and that should still have "none" alignment stored for it?

e-kolpakov commented:
@itsjeyd Thank you for updating the sandbox! For testing existing answers - I believe sandbox 3 should be free and not updated to this version, so you could use it to post answers, deploy this PR and check this feature.

On the other hand it might be esaier to test this in devstack :)

itsjeyd commented:
@e-kolpakov Verified that last item on devstack; it works as expected. The only thing I noticed was that for problems in assessment mode, decoy items got returned to the bank after updating to the code in this branch:

1. Add default DnDv2 problem to unit and set mode to "assessment".
2. Drag items to board in the LMS, including decoy item. Result: All items on board, none in item bank.
3. Switch from `master` to `ekolpakov/non-overlapping-placement`.
4. Reload problem in LMS. Result: Normal items on board (correctly centered in their zones), decoy item back in item bank.

If the changes from this PR are going to go live on edx.org together with the main implementation of assessment mode, I don't expect this to be a problem. But if the changes from this PR land at a later date, the behavior described above should probably be fixed now.

---

One more thing: Please update the README (shouldn't say that item alignment is optional, and shouldn't mention "No alignment is the default", etc.)

itsjeyd commented:
:+1: once last remaining comments have been addressed ([1](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#discussion_r74623925), [2](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#discussion_r74624004), [3](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#issuecomment-239749633)) and build is green.

- [x] I tested this: Features work as advertised in description
- [x] I read through the code
- [ ] Includes documentation -- **not yet** (will update this comment once that's done)

e-kolpakov commented:
@itsjeyd 

> The only thing I noticed was that for problems in assessment mode, decoy items got returned to the bank after updating to the code in this branch

That's because this branch is not based on #87, so decoy items are not saved when dropped onto a zone. Also, when I switched to this branch the decoy item stayed where they were - it was returned to bank only after I reset the problem and tried dropping it again.

So, I don't think it will be a problem - this PR have some value on it's own (as it affects standard mode), but since it is not based on #87 assessment mode is still non-functional here. So it shouldn't be deployed without #87.

e-kolpakov commented:
@itsjeyd rebased from f18278a9e87bd700c57f9a547d279403f2341352

itsjeyd commented:
@e-kolpakov Thanks for the documentation update, I adjusted my earlier [comment](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#issuecomment-239750103).

e-kolpakov commented:
@cahrens, @staubina, @cptvitamin, @sstack22 this PR is ready for your review. 

CI build is red, but only [flaky test failures ](https://travis-ci.org/edx-solutions/xblock-drag-and-drop-v2/builds/152385035) on the last build.

cahrens commented on README.md:
```diff
@@ -114,12 +114,11 @@ whether or not to display borders outlining the zones. It is possible
 to right across the zone. Right alignment causes the items to be placed from
-right to left across the zone. Center alignment places items from top to bottom
-along the center of the zone. If left, right, or center alignment is chosen,
+right to left across the zone. If left, right, or center alignment is chosen,
```
Reword "**If** left, right, or center...". One of those has to be chosen now.

cahrens commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -110,6 +113,13 @@ class DragAndDropBlock(XBlock, XBlockWithSettingsMixin, ThemableXBlockMixin):
         default="",
     )
 
+    max_items_per_zone = Integer(
```
It feels odd that there is a single value for all zones in the problem. One can imagine that some zones are much larger than other zones. But product has decided that this is a valuable option to have?

e-kolpakov replied:
@cahrens I've talked [about that]https://openedx.atlassian.net/browse/SOL-1979?focusedCommentId=206512&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-206512) and there were no objections to implementing this approach. So I would assume product is fine with single setting for all zones. 

@sstack22 - any comments? Just FYI, I'll be away starting Friday and I would really love to land all my PRs (or at least this and #93) before that; implementing per zone max items would make it problematic :)

cahrens commented:
Display of errors (in Studio) needs UX improvements. Also keyboard focus needs to move to the error notification area.

![image](https://cloud.githubusercontent.com/assets/484484/17677323/4ac98d04-6300-11e6-8363-c5cd985197df.png)

cahrens commented on drag_and_drop_v2/public/js/drag_and_drop_edit.js:
```diff
@@ -524,8 +521,13 @@ function DragAndDropEditBlock(runtime, element, params) {
-                                $('.xblock-editor-error-message', element)
-                                    .html(gettext('Error: ') + response.message);
+                                var errorMsg = $("<p></p>").text(gettext('Error: '));
+                                $('.xblock-editor-error-message', element).append(errorMsg);
```
Note that previous error messages are never cleared out (save with error, save again with same or different error).

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -133,6 +133,9 @@
                       {{ help_texts.item_text_color }}
                     </div>
+                    <label class="h3" for="max-items-per-zone" title="{{ help_texts.max_items_per_zone }}">{% trans "Maximum items per zone" %}</label>
+                    <input name="max-items-per-zone" id="max-items-per-zone" type="number" min="1" step="1" value="{{ self.max_items_per_zone|unlocalize }}" />
```
Nest input inside of label so you don't need to use ids.

e-kolpakov commented:
@cahrens Unfortunately I'm no UX expert, so the best I can do to improve error UX is this:

![image](https://cloud.githubusercontent.com/assets/3330048/17700208/7408778e-63cd-11e6-8d04-32b39179a9e4.png)

e-kolpakov commented:
@cahrens, @staubina, @cptvitamin, @sstack22 Same here as on [other PR](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/93#issuecomment-240344038) - I would love to finish it before my vacation, so I would be very grateful if you could help me with it.

@cptvitamin this can be a part of [SOL-1978](https://openedx.atlassian.net/browse/SOL-1978), so I believe we can go without you reviewing/approving this PR right now.

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -116,6 +116,7 @@
                 <h3>{% trans "Items" %}</h3>
             </header>
             <section class="tab-content">
+                <div class="xblock-editor-error-messages" aria-live="polite" tabindex="0"></div>
```
tabindex should be -1, not 0

cahrens replied:
Doesn't this div need to be present on all the tabs? Right now it is just within "Items".

e-kolpakov replied:
@cahrens Fixed aria-live and tabindex.

As for showing this div on all tabs - it only shows error feedback from backend, and this tab is the only one actually sending the feedback. Your suggestion totally makes sense, but it is not necessary now, so I would keep it where it is needed only.

To be honest, I don't like the editor much, so we might want to completely overhaul it (both design and UX) later. But that's my point of view :)

cahrens replied:
That feels very icky-- only showing it on one tab because that's where it is currently used. The JavaScript method that you call to show the notification is generic, correct? Isn't there a general place where it can be shown (perhaps above the "Items" label)? It felt weird anyway that the error was injected into that location anyway. It's great if the error is inline with the control that caused it, but that isn't the case; it was just put into a place that felt a bit arbitrary anyway.

e-kolpakov replied:
@cahrens No, it's not generic - those method collects data from all the "tabs" and sends it; each individual tab does not send anything. The validation on previous tabs are client-side only (btw, I don't like it as well - it is relatively easy to workaround that).

Putting the error next to control is a sensible suggestion, and actually I evaluated that option first; unfortunately, it would require somewhat involving overhaul of editor, so, taking into account deadlines, my vacation and other things I've decided that current solution would be the best I could reliably deliver. 

To explain this decision a bit further, in this particular case those errors are caused by groups of checkboxes that are part of "item" block; there are always more than one such "item" blocks; and so far they are indistinguishable from each other. So, to implement that suggestion I would have to assign some identifiers to the fields, or at least "item" blocks + make sure backend validation returns errors associated with those identifiers somehow.

Since this task is not about improving author experience, I decided to stop right there and follow a simpler approach.

Also, I really love how me being proactive about some possible problems brings just more and more feedback, requesting my little improvement not even mentioned in the task description/acceptance criteria (no one thought about that) to be absolutely bulletproof :)

Also, I've just finished a11y course training, and one of the recommendations was to put errors summary *in the beginning* of the form, so I'll put it there - in the beginning of the form. Visually it wouldn't make any difference though.

Should we bring someone from UX team maybe?

bradenmacdonald replied:
I definitely want to avoid re-working the XBlock editor substantially for this PR if we can help it.

If the UX for this particular validation is a problem, I'd rather leave out this warning. Besides, I think there's a bug in the current logic: if I have 10 items and two zones and I limit each zone to 5 items but allow each item to be placed in any zone, then it will raise an error, though that's a perfectly valid problem. Validating this in a way that accounts for such problems is likely going to be overly complex.

cahrens commented:
Inconsistent spacing between label and input. See how other items have some padding between label and input field.

![image](https://cloud.githubusercontent.com/assets/484484/17743239/fb8e791c-6470-11e6-981e-e1b339dd74c0.png)

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -133,6 +134,12 @@
                     </div>
+                    <label class="h3" title="{{ help_texts.max_items_per_zone }}">
+                        {% trans "Maximum items per zone" %}
+                        <br/>
```
Why br? The styling should be one with CSS.

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -133,6 +134,12 @@
+                    <label class="h3" title="{{ help_texts.max_items_per_zone }}">
+                        {% trans "Maximum items per zone" %}
+                        <br/>
+                        <input name="max-items-per-zone" id="max-items-per-zone" type="number" min="1" step="1" value="{{ self.max_items_per_zone|unlocalize }}" />
```
It is better to use classes instead of IDs.

cahrens commented on drag_and_drop_v2/templates/html/drag_and_drop_edit.html:
```diff
@@ -133,6 +134,12 @@
+                        <br/>
+                        <input name="max-items-per-zone" id="max-items-per-zone" type="number" min="1" step="1" value="{{ self.max_items_per_zone|unlocalize }}" />
+                    </label>
+                    <span class="sr">{{ help_texts.max_items_per_zone }}</span>
```
Don't think you need to duplicate the title text in a screen reader span (which isn't linked to the input anyway).

cahrens commented:
@e-kolpakov I don't have time to review the migration and test code (much of which appeared to just move around, making it difficult for me to tell what is new). I am done reviewing the rest of the PR.

cahrens commented:
@e-kolpakov If the errors are collected across all tabs, why are they shown on one specific tab? Because it is the last one displayed? But then the errors may not be related to "Items" at all, so placing the div within "Items" doesn't make sense to me.

e-kolpakov commented:
@cahrens 

> If the errors are collected across all tabs, why are they shown on one specific tab? Because it is the last one displayed? But then the errors may not be related to "Items" at all, so placing the div within "Items" doesn't make sense to me.

Proper solution to this problem will require a significant rework of the editor - for example there are no way to go back to previous tab.

Also, only errors from "Items" tab are validated on server, so unless more validation checks are added, error messages will only contain errors from this tab.

bradenmacdonald commented:
@e-kolpakov There is already an existing UX for displaying errors built in to studio, which we should use rather than making something idiosyncratic. I believe that all you need to do to show an error message is:

```javascript
runtime.notify('error', {title: gettext("Error title here"), message: gettext("Error message here")});
```

You can use this for client-side validation or you can follow [this pattern](https://github.com/edx/xblock-utils/blob/24f20b8/xblockutils/public/studio_edit.js#L116-L139) to use it for server-side validation.

e-kolpakov commented:
@bradenmacdonald That's a questionable improvement:

![image](https://cloud.githubusercontent.com/assets/3330048/17782009/4d117f78-657a-11e6-9bff-11c7cc77614e.png)

* It is not very well suited to showing multiple error messages, like what we might have here.
* It is much farther from the controls that caused the error than current solution
* It does not address most of the feedback I've got:
** Only shown on "Items" tab
** Does not highlight controls caused the error in any way.

So, looks like removing the validation is a best course of action, especially taking into account [this comment](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#discussion_r75341535) (good catch btw).

cahrens commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -306,6 +317,14 @@ def studio_submit(self, submissions, suffix=''):
         """
         Handles studio save.
         """
+        errors = self._validate_submissions(submissions)
```
I thought the validation was being removed.

haikuginger commented:
@cahrens, validation should now be more thoroughly removed.

itsjeyd commented on drag_and_drop_v2/utils.py:
```diff
@@ -78,3 +78,139 @@ def not_placed(number, ngettext=ngettext_fallback):
+    @classmethod
+    def _item_state_v1p5_to_v2p1(cls, item):
+        """
+        Migrates item_state from v1.5 to v2.5
```
@haikuginger Following up on https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#discussion-diff-74210508, this should be `v2.0` to `v.2.1`.

itsjeyd commented on README.md:
```diff
@@ -122,15 +122,13 @@ whether or not to display borders outlining the zones. It is possible
-potentially, overlap the zones below.
+right to left across the zone. Items dropped in a zone will not overlap, 
+but if the zone is not made large enough for all its items, they will overflow the bottom 
+of the zone, and potentially, overlap the zones below.
```
@haikuginger Remove comma after "potentially".

itsjeyd commented on README.md:
```diff
@@ -143,6 +141,9 @@ correct zone, while the error feedback is shown when dropping the item on an
 incorrect drop zone.  In assessment mode, the success and error feedback texts
 are not used.
 
+"Maximum items per Zone" setting controls how much items can be dropped into a 
```
@haikuginger "how many items"

Also, it would make sense to move this paragraph to the end of this section: The paragraphs above and below talk about per-item settings, while this one describes a setting that applies to all items.

itsjeyd commented on drag_and_drop_v2/public/css/drag_and_drop_edit.css:
```diff
@@ -371,3 +371,12 @@
+.xblock--drag-and-drop--editor .xblock-editor-error-messages ul li {
+    list-style-type: disc;
+    margin-left: 20px;
+}
```
@haikuginger The `xblock-editor-error-messages` class doesn't seem to be used anywhere; I think you can remove the rules targeting it.

itsjeyd commented:
@haikuginger I made another pass over the PR to double-check whether all comments had been addressed. There's a few things that slipped through or should be confirmed after rebasing:

* Inconsistent spacing between label and input ("Maximum item per zone setting"), see https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#issuecomment-240456637. AFAICT, this hasn't been addressed in any of the updates that were pushed after Christina posted the comment.
* Help text for "Maximum items per zone setting" should be displayed in Studio, see https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#issuecomment-239389310. This was deferred to the story for improving help texts ([OC-1778](https://tasks.opencraft.com/browse/OC-1778)) in a follow-up comment (https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#issuecomment-239605793), but the PR for that story (#88) was merged in the meantime, so we'll need to take care of it in this context.
* Decoy items should stay on board after reload, see https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#issuecomment-239749633. As per https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92#issuecomment-239792271, this should work once after rebasing, which will pull in changes from #87.

I will do a final round of manual testing when you're done rebasing and addressing these items, and the sandbox is up-to-date.

haikuginger commented:
Rebased from 123235a9431f43f23b65188e44ef5c8969ff3e82; comments addressed. @itsjeyd, I'm reprovisioning the sandbox right now.

haikuginger commented:
@itsjeyd, actually, hold off a bit - there's some stuff that automerge didn't catch.

haikuginger commented:
@itsjeyd, sorry for the delay. This is set up in the sandbox and ready for review.

itsjeyd commented on drag_and_drop_v2/utils.py:
```diff
@@ -78,3 +78,139 @@ def not_placed(number, ngettext=ngettext_fallback):
+        """
+        attributes_to_remove = ['x_percent', 'y_percent', 'left', 'top', 'absolute']
+        for attribute in attributes_to_remove:
+            del[attribute]
```
@haikuginger This should be `del item[attribute]`, otherwise `item` will be returned unmodified.

mtyaka commented:
@itsjeyd @haikuginger I just noticed the state migration related changes in this PR - #98 is somewhat related.

haikuginger commented:
Well spotted, @itsjeyd. I've switched that to a `pop` to avoid any possible KeyErrors.

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -127,6 +129,13 @@ class DragAndDropBlock(XBlock, XBlockWithSettingsMixin, ThemableXBlockMixin):
 
+    max_items_per_zone = Integer(
+        display_name=_("Maximum items per zone"),
+        help=_("This setting limits the number of items that can be dropped into a single zone"),
```
@haikuginger Please add a period to make this consistent with help messages for other fields, and update the translation files.

itsjeyd commented:
@haikuginger I tested on the sandbox, and most everything is working as expected. The only issue I found was:

* Can't save zone alignment: When opening a DnDv2 instance in Studio after changing the alignment of a zone and submitting the changes, the value is back to "center".

I also had a look at the commits you've pushed so far. They look good (although the commit message of https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92/commits/074a8571c69a7515a80d25f1a18249971036827e is fairly cryptic ;))

haikuginger commented:
@itsjeyd, those changes are deployed. We seem to be down to pretty minor stuff, so if it's all right, I'll wait to update the sandbox until we've got a 👍.

itsjeyd commented:
Rebased from https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/92/commits/a58f72d3835e6825cdaf523effa6f4f525f3fb12.

itsjeyd commented:
@haikuginger Thanks for the updates, they fix the problem. In my capacity as sprint firefighter I took the liberty to push this forward some more: I rebased the branch, cleaned up the test code post-rebase, deployed the updated code to the sandbox, and tested everything again according to the PR description (including behavior for legacy zones with "none" alignment -- this I verified in my local dev env, since it requires working with different branches).

This is ready for upstream review now. :+1: 

- [x] I tested this: Features work as advertised in description
- [x] I read through the code
- [x] Includes documentation

cahrens commented:
@itsjeyd Is this PR ready for TNL re-review?

itsjeyd commented:
Hi @cahrens, yes, this PR is ready for TNL re-review. Thanks!

cahrens commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -315,12 +325,34 @@ def studio_submit(self, submissions, suffix=''):
         self.weight = float(submissions['weight'])
         self.item_background_color = submissions['item_background_color']
         self.item_text_color = submissions['item_text_color']
+        # Possible ValueError should be catched in _validate_submissions
```
This comment is no longer relevant, correct?

cahrens commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -315,12 +325,34 @@ def studio_submit(self, submissions, suffix=''):
+        """
+        Parses Max items per zone value coming from editor.
+
+        Does not catch ValueError intentionally to allow validation to capture this problem
```
Also not sure what you want to do about this.

cahrens commented:
Nit: the following scenario feels a bit odd:

1) Set maximum items per zone to 2.
2) Drag 2 items to the zone.
3) Pick up one of the items and release it (back into the same zone).
4) You get an error message about not adding more items to the zone.

I wouldn't expect an error message because I'm not adding new items to the zone, just "moving them around".

e-kolpakov commented:
@cahrens Notes about comments addressed, showing error message when dropping an item to the same zone is fixed, sandbox is updated

sstack22 commented:
I've finished my review, just one small comment:

- I believe the help text for alignment needs to be updated: "Default is no alignment (items stay exactly where the user drops them)." This is no longer true, correct?

cahrens commented:
👍 on review feedback commits. I still have not reviewed the migration and test code, but I trust your internal reviews. 

Please do see @sstack22's comment above though about documentation.

e-kolpakov commented:
@cahrens @sstack22 rebased from 4beaba2 and fixed help text for zone alignment. Updating now.

sstack22 commented:
@e-kolpakov - great, confirmed the fix! 👍

e-kolpakov commented:
Squashed commits - previous revision fabb8f608b3cae52948b3195bb7a9c9b19b6033b
